### PR TITLE
Scikit-Learn compatibility

### LINKED ISSUE #148: Make AutoML scikit-learn compatible
I think it would be interesting to add a feature to export the best model with a scikit-learn wrapper. This would allow integrating the best AutoML model into a scikit-learn workflow.
I think most of the models that AutoML uses are already from scikit-learn, and those who aren't do provide scikit-learn wrappers, so I think it would be easy to implement.
Is there anything that makes this feature 'impossible' to implement?

### PR DESCRIPTION
**Pull request by spamz23**
Fixes #148 
To give AutoML scikit-learn compatibility I did the following:

- Re-wrote AutoML class, making it `_AutoML `that inherits from scikit-learn BaseEstimator, and ABC (making it abstract)
- Wrote a new AutoML class (inherits from `_AutoML `), only exposing the `fit()` ,`predict()`, `predict_proba()`, `score()` methods to the user. This is a scikit-learn standart.
- Changed the way the parameters are passed in the constructor, making it cleaner.
- Improved the documentation (numpydoc)

### CODE REVIEW

**Comment by spamz23** [comment]
@pplonski travis build failed, because I didn't update the tests. A lot of tests use `set_advanced` method. However, I removed it from `AutoML`. All the params are now passed in constructor, and can be changed via `set_params(**kwargs)`

**Comment by pplonski** [comment]
@spamz23 thank you for the contribution!

The code looks fantastic!

I have a few concerns:
- If I see a PR with failing tests - means that changes are suspected.
- There are a lot of changes in the code. It is hard to check them all. I checked the errors from Travis and found a few bugs that are added with this code refactoring. 

For example:

https://github.com/spamz23/mljar-supervised/blob/66cccd0cb92bc8493426517ae15e5cec097f257d/supervised/automl.py#L75

The variable `self._results_path` is set to `None` just before the reading from it.

Next example, (from loading functionality):

https://github.com/spamz23/mljar-supervised/blob/66cccd0cb92bc8493426517ae15e5cec097f257d/supervised/automl.py#L800

reading from variable `'self.result_path` , of course, it is a typo (or maybe not, there are two variables from what I see `self._results_path` and `self.results_path`).

- there is a change in validation params `X_train_path` -> `X_path`. I like the shorter form. But this change will make older models not backward compatible (in case of stopped training and then restored). Probably it is a very small issue. But I didn't spend a lot of time in code analysis, maybe there are others. There are not many tests to cover all cases and with the current state, almost 10% of tests are failing. 

I don't know what to do next. I really like your code changes. The whole code improved and was easier for me to read. But I'm afraid of new bugs that are introduced in this change.

**Comment by spamz23** [comment]
Hey @pplonski ! Yes there are some bugs on the PR. I need to fix them. Some on the fails are still because of tests that need to be adapted, which I'll take care as well.
The change in `X_train_path` to `X_path` is to provide consistency with the new names on `fit()`. The should should abstract from the "type" of data. Is up to the user to decide if it fits training data, validation data, or entire dataset (for production environment).
The case of backward compatibility didn't cross my mind tbh.
Wait for me to re-write the old tests, write new ones to extensively cover the new code, and fix the small bugs.

**Comment by pplonski** [comment]
@spamz23 all right! :clap: I'm very impressed with your code and I would definitely like to see it in the master.

The package is still in the early stages of development, but it's important to keep backward compatibility in mind. For `X_path` let's keep the new one.

**Comment by spamz23** [comment]
@pplonski can you review this?

**Comment by pplonski** [comment]
hey @spamz23! Fantastic contribution, I have a feeling that you rewritten the whole package. :+1:

I will review the changes in the morning. Please fill the Contributor License Agreement https://github.com/mljar/mljar-supervised/blob/master/docs/contributing.md

**Comment by spamz23** [comment]
@pplonski done

File: supervised/automl.py [hunk 1]
```diff
@@ -866,74 +461,124 @@ def load_progress(self):
...
+        # Must pass copies otherwise, X and y get lost
+        self._save_data(deepcopy(X), deepcopy(y))
+
+        self.verbose_print(f"AutoML current directory: {self._results_path}")
+        self.verbose_print(f"AutoML current task: {self._ml_task}")
+        self.verbose_print(f"AutoML will use algorithms : {self._algorithms}")
+        self.verbose_print(f"AutoML will optimize for metric : {self._eval_metric}")
+
+        try:
+            self.load_progress()
```
Comment (pplonski) [inline supervised/automl.py:549]: There is a missing check if `AutoML` was already fitted.

File: supervised/automl.py [hunk 2]
```diff
@@ -408,13 +127,16 @@ def load(self):
...
+            with open(os.path.join(path, "best_model.txt"), "r") as fin:
                 best_model_name = fin.read()
 
             self._best_model = models_map[best_model_name]
 
-            data_info_path = os.path.join(self._results_path, "data_info.json")
+            data_info_path = os.path.join(path, "data_info.json")
             self._data_info = json.load(open(data_info_path))
+            self.n_features = self._data_info["n_features"]
+            self.n_classes = self._data_info["n_classes"]
```
Comment (pplonski) [inline supervised/automl.py:138]: `n_classes` will be confusing for regression tasks.

File: supervised/automl.py [hunk 3]
```diff
@@ -866,74 +461,124 @@ def load_progress(self):
...
-                X_train, y_train, X_validation, y_validation
-            )
-            self._save_data(X_train, y_train, X_validation, y_validation)
-            self._set_algorithms()
-            self._set_metric()
-
-            if self._ml_task in [BINARY_CLASSIFICATION, MULTICLASS_CLASSIFICATION]:
-                self._check_imbalanced(y_train)
+                EDA.compute(
+                    deepcopy(X), deepcopy(y), os.path.join(self._results_path, "EDA")
```
Comment (pplonski) [inline supervised/automl.py:558]: We don't need to copy the data to make EDA. There are no changes in the data in EDA going on.

File: supervised/automl.py [hunk 4]
```diff
@@ -1087,11 +746,472 @@ def predict(self, X):
...
+        # If classification task the result is in column 'label'
+        # Need to drop `label` column.
+        return self._base_predict(X).drop(["label"], axis=1).to_numpy()
+
+    def _score(self, X, y=None):
+        # y default must be None for scikit-learn compatibility
+
+        # Check if y is None
+        if y is None:
+            raise AutoML("y must be specified.")
```
Comment (pplonski) [inline supervised/automl.py:784]: Probably should be AutoMLException

File: supervised/automl.py [hunk 5]
```diff
@@ -1113,3 +1233,331 @@ def from_json(self, json_data):
...
+
+    def __init__(
+        self,
+        mode="Explain",
+        ml_task="auto",
+        tuning_mode="Normal",
+        results_path=None,
+        total_time_limit=30 * 60,
+        model_time_limit=None,
+        algorithms="auto",
```
Comment (pplonski) [inline supervised/automl.py:1453]: I kept all `auto` parameters in `kwargs` because  I want to have a simple constructor for `AutoML`. 

The inexperienced user will find it scary to see so many parameters in the constructor.

File: supervised/automl.py [hunk 6]
```diff
@@ -1087,11 +746,472 @@ def predict(self, X):
...
+        # Check is task type is correct
+        if self._ml_task == REGRESSION:
+            raise AutoMLException(
+                f"Method `predict_proba()` can only be used when in classification tasks. Current task: '{self._ml_task}'."
+            )
+
+        # Make and return predictions
+        # If classification task the result is in column 'label'
+        # Need to drop `label` column.
+        return self._base_predict(X).drop(["label"], axis=1).to_numpy()
```
Comment (pplonski) [inline supervised/automl.py:777]: In case of large ensembles, if user will need labels and probabilities, then he will need to run both `predict` and `predict_proba`. I think there should be `predict_all` method to predict probabilities and labels at once.

File: tests/tests_automl/test_automl.py [hunk 7]
```diff
@@ -5,61 +5,153 @@
...
+        X, y = datasets.make_classification(n_samples=30)
+        # apply PCA to X
+        new_X = PCA(random_state=0).fit_transform(X)
+        # Create default model
+        default_model = AutoML(algorithms=["Linear"], random_state=0)
+        # Fit default model with transformed X and y, and predict transformed X
+        y_pred_default = default_model.fit(new_X, y).predict(new_X)
+
+        # Create pipeline with PCA and AutoML
+        pipeline = make_pipeline(
```
Comment (pplonski) [inline tests/tests_automl/test_automl.py:128]: Nice!

File: tests/tests_automl/test_targets.py [hunk 8]
```diff
@@ -34,13 +34,12 @@ def test_bin_class_01(self):
             algorithms=["Xgboost"],
             train_ensemble=False,
             explain_level=0,
+            start_random_models=1,
         )
-        automl.set_advanced(start_random_models=1)
         automl.fit(X, y)
         pred = automl.predict(X)
-        for col in ["prediction_0", "prediction_1", "label"]:
```
Comment (pplonski) [inline tests/tests_automl/test_targets.py:41]: Maybe there should be `predict_all` method?

**Review by pplonski — CHANGES_REQUESTED** [review]
Please see the comments in the code.

File: supervised/automl.py [hunk 9]
```diff
@@ -7,6 +7,14 @@
 import pandas as pd
 import logging
 from tabulate import tabulate
+from abc import ABC
+from copy import deepcopy
+
+from varname import nameof
```
[the remark this answers is not in the document]
Comment (spamz23) [reply supervised/automl.py:13]: If don't want `varname`, instead of `nameof`, name of functions can be passed manually.

File: supervised/automl.py (hunk 2 above)
Comment (spamz23) [reply supervised/automl.py:138]: Should I add a check to only add `n_classes` for classification tasks?

File: supervised/automl.py (hunk 1 above)
Comment (spamz23) [reply supervised/automl.py:549]: Part of the scikit-learn ideology is that a model can be fitted `n` times. Any `fit()` call should be independent from the previous calls.

File: supervised/automl.py (hunk 6 above)
Comment (spamz23) [reply supervised/automl.py:777]: In which format `predict_all` would return the data? The original Pandas dataframe?

File: supervised/automl.py (hunk 4 above)
Comment (spamz23) [reply supervised/automl.py:784]: Yes! Thanks for noticing the typo. Gonna add a test case for when `y` is `None`

File: supervised/automl.py (hunk 5 above)
Comment (spamz23) [reply supervised/automl.py:1453]: I get your point. But `AutoML` can also work with no params provided. So a inexperienced user can just `model = AutoML()`

File: supervised/automl.py (hunk 3 above)
Comment (spamz23) [reply supervised/automl.py:558]: Okok! Will fix it

File: supervised/automl.py [hunk 10]
```diff
@@ -7,6 +7,14 @@
...
+
+from sklearn.base import BaseEstimator
+from sklearn.utils.validation import check_array
+from sklearn.metrics import r2_score, accuracy_score
```
[the remark this answers is not in the document]
Comment (spamz23) [reply supervised/automl.py:17]: These metrics are used in `score()`. This method must return a goodness of `fit` measure (higher the better). I choose these metrics because they are the most common ones.

File: supervised/automl.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (pplonski) [reply supervised/automl.py:13]: Yep, I think it will be better to just go with name of the function. Why have you used `nameof`? What is the advantage?

File: supervised/automl.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (pplonski) [reply supervised/automl.py:17]: OK, let's leave it as it is. When implementing user set `eval_metric` we can improve on this.

File: supervised/automl.py (hunk 1 above)
Comment (pplonski) [reply supervised/automl.py:549]: But in the case of AutoML the model training can be many hours. The training can be interrupted and then restored. In this case, I won't keep the scikit-learn ideology.

File: supervised/automl.py (hunk 2 above)
Comment (pplonski) [reply supervised/automl.py:138]: I think yes, it is important to not bother the user with unnecessary things.

File: supervised/automl.py (hunk 3 above)
Comment (pplonski) [reply supervised/automl.py:558]: If I remember, the same was for saving (somewhere in the code). Could you fix saving as well.

File: supervised/automl.py (hunk 4 above)
Comment (pplonski) [reply supervised/automl.py:784]: :+1:

File: supervised/automl.py (hunk 6 above)
Comment (pplonski) [reply supervised/automl.py:777]: yes, pandas data frame

Comment (spamz23) [reply supervised/automl.py:777]: The pandas dataframe returned by `_base_predict()` contains `prediction_0`, ..., `prediction_n`, `label`. Is this fine?
The method would be:
```
def _predict_all(self, X):
        # Check is task type is correct
        if self._ml_task == REGRESSION:
            raise AutoMLException(
                f"Method `predict_all()` can only be used when in classification tasks. Current task: '{self._ml_task}'."
            )

        # Make and return predictions
        return self._base_predict(X)
```

File: supervised/automl.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (spamz23) [reply supervised/automl.py:13]: I think there is no real "winning" from this package. It just seemed cleaner to use it.

**Comment by spamz23** [comment]
Ups, broke the tests. Fixing it

File: requirements.txt [hunk 11]
```diff
@@ -16,4 +16,5 @@ matplotlib>=3.2.2
 dtreeviz==0.8.2
 shap==0.35.0
 seaborn==0.10.1
-wordcloud==1.7.0
\ No newline at end of file
+wordcloud==1.7.0
+python-varname==0.4.0
```
[the remark this answers is not in the document]
Comment (spamz23) [reply requirements.txt]: Deleted it

File: tests/tests_automl/test_targets.py (hunk 8 above)
Comment (spamz23) [reply tests/tests_automl/test_targets.py:41]: Implemented

File: supervised/automl.py (hunk 1 above)
Comment (spamz23) [reply supervised/automl.py:547]: Right now `fit()` can be called multiple times, and it will continue previous training.

File: supervised/automl.py (hunk 3 above)
Comment (spamz23) [reply supervised/automl.py]: Removing `deepcopy ` on saving causes EDA to crash, as `X `and `y ` get empty after the `save` method. Don't really know why `X ` and `y` get empty when passed to save method. Have to look deeper into that, but for now, `deepcopy `solves the issue.

File: supervised/automl.py (hunk 2 above)
Comment (spamz23) [reply supervised/automl.py]: Fixed this

File: supervised/automl.py (hunk 1 above)
Comment (spamz23) [reply supervised/automl.py:554]: Fixed this. For now `fit` can only be called once

**Review by pplonski — APPROVED** [review]

**Comment by pplonski** [comment]
Thank you @spamz23 amazing contribution! :tada:
